File: csharp/src/Containers/GPMemoryManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EPAM.Deltix.Containers{
	/// <summary>
	/// Geometric progression memory manager.
	/// </summary>
	/// <typeparam name="T">Thpe of object to manage.</typeparam>
	public class GPMemoryManager<T> : IMemoryManager<T> where T : new()
	{
		T[] freeObjects;
		Int32 numberOfFreeObjects = 0;
		Int32 totalObjectsCount = 0;
		Double coefficient;

		/// <summary>
		/// Default maximal value of number object increment.
		/// </summary>
		public Int32 MaxIncrement
		{
			get; set;
		} = 1000000;

		/// <summary>
		/// Creates new geometric progression memory manager.
		/// </summary>
		/// <param name="capacity">Initial pool capacity.</param>
		/// <param name="coefficient">Scale factor of geometric progression.</param>
		public GPMemoryManager(Int32 capacity, Double coefficient)
		{
			if (coefficient <= 1.0)
				throw new ArgumentException("coefficient");

			this.coefficient = coefficient;
			freeObjects = new T[capacity];
			totalObjectsCount = capacity;
			for (int i = 0; i < capacity; ++i)
				freeObjects[i] = new T();
		}

		/// <summary>
		/// Creates new geometric progression memory manager.
		/// </summary>
		public GPMemoryManager()
			: this(0x10, 2)
		{
		}

		/// <summary>
		/// Reserve new instance of an object.
		/// </summary>
		/// <returns>Instance of an object</returns>
		public T New()
		{
			if (numberOfFreeObjects == 0)
			{
				Int32 length = freeObjects.Length;
				Int32 newLength = Math.Min(MaxIncrement, (Int32)(freeObjects.Length * coefficient));
				freeObjects = new T[newLength];
				numberOfFreeObjects = newLength;

				for (int i = 0; i < numberOfFreeObjects; ++i)
					freeObjects[i] = new T();

			}
			numberOfFreeObjects--;
			return freeObjects[numberOfFreeObjects];
		}

		/// <summary>
		/// Free instance of an object and return it to the pool.
		/// </summary>
		/// <param name="obj">Instance of an object to free.</param>
		public void Delete(T obj)
		{
			if (numberOfFreeObjects == freeObjects.Length)
			{
				Array.Resize(ref freeObjects, freeObjects.Length << 1);
			}
			freeObjects[numberOfFreeObjects] = obj;
			numberOfFreeObjects++;
		}

		/// <summary>
		/// Number of free objects in pool.
		/// </summary>
		public Int32 FreeObjectsCount { get { return numberOfFreeObjects; } }

		/// <summary>
		/// Total number of objects in pool free + reserverd.
		/// </summary>
		public Int32 TotalObjectsCount { get { return freeObjects.Length; } }
	}
}

File: java/src/main/java/com/epam/deltix/containers/GPMemoryManager.java
package com.epam.deltix.containers;

import com.epam.deltix.containers.interfaces.MemoryManager;

/**
 * Memory pool of objects with T type interface. It grows in geometric progression.
 * @param <T> Type of objects
 */
@SuppressWarnings("unused")
public class GPMemoryManager<T> implements MemoryManager<T> {
    /**
     * The value of initial capacity that will be used if not provided.
     */
    @SuppressWarnings("WeakerAccess")
    public static final int DEFAULT_INITIAL_CAPACITY = 4;

    /**
     * Default maximal value of number object increment
     */
    public static final int DEFAULT_MAXIMAL_INCREMENT = 1000000;


    private int maxIncrement = 1000000;

    /**
     * The value of growth factor that will be used if not provided.
     */
    @SuppressWarnings("WeakerAccess")
    public static final double DEFAULT_GROWTH_FACTOR = 2.0;

    private final Creator<T> creator;
    private final int initialCapacity;
    private final double growthFactor;

    private T[] freeObjects = null;
    private int numberOfFreeObjects = 0;

    /**
     * Create new instance of MemoryManager.
     * @param creator Objects creator
     * @param initialCapacity Initial capacity of pool
     * @param growthFactor Growth geometric progression factor.
     */
    @SuppressWarnings("WeakerAccess")
    public GPMemoryManager(MemoryManager.Creator<T> creator, int initialCapacity, double growthFactor) {
        if (creator == null)
            throw new IllegalArgumentException("Argument 'creator' cannot be null.");
        if (growthFactor <= 1.0)
            throw new IllegalArgumentException("Argument 'growthFactor' should be greater than 1.0.");
        if (initialCapacity < 0)
            throw new IllegalArgumentException("Argument 'initialCapacity' cannot be negative.");

        this.creator = creator;
        this.initialCapacity = initialCapacity;
        this.growthFactor = growthFactor;
    }

    /**
     * Create new instance of MemoryManager.
     * @param creator Objects creator.
     */
    @SuppressWarnings("unused")
    public GPMemoryManager(MemoryManager.Creator<T> creator) {
        this(creator, DEFAULT_INITIAL_CAPACITY, DEFAULT_GROWTH_FACTOR);
    }

    /**
     * Set maximal increment to new value.
     * @param maxIncrement Maximal increment.
     */
    public void setMaxIncrement(int maxIncrement) {
        this.maxIncrement = maxIncrement;
    }

    /**
     * Create new instance of MemoryManager.
     * @param clazz Objects class.
     * @param initialCapacity Initial capacity.
     * @param growthFactor Growth geometric progression factor.
     */
    @SuppressWarnings("WeakerAccess")
    public GPMemoryManager(Class<T> clazz, int initialCapacity, double growthFactor) {
        this(new ReflectionObjectCreator<>(clazz), initialCapacity, growthFactor);
    }

    /**
     * Create new instance of MemoryManager.
     * @param clazz Objects class.
     */
    @SuppressWarnings("WeakerAccess")
    public GPMemoryManager(Class<T> clazz) {
        this(clazz, DEFAULT_INITIAL_CAPACITY, DEFAULT_GROWTH_FACTOR);
    }

    /**
     * Return number of free objects in pool
     * @return Number of free objects in pool
     */
    @Override
    public int getFreeObjectsCount() {
        return numberOfFreeObjects;
    }

    /**
     * Return total number of objects from pool
     * @return Total number of objects from pool
     */
    @Override
    public int getTotalObjectsCount() {
        return freeObjects.length;
    }

    /**
     * Return new free object from pool
     * @return New free object from pool
     */
    @Override
    public T getNew() {
        if (numberOfFreeObjects == 0) {
            final int capacity = freeObjects != null ? freeObjects.length : 0;
            final int newCapacity = Math.min(maxIncrement, capacity > 0 ? (int) (capacity * growthFactor) : initialCapacity);
            freeObjects = creator.create(newCapacity);
            numberOfFreeObjects = newCapacity;
            for (int i = 0; i < numberOfFreeObjects; i += 1)
                freeObjects[i] = creator.create();
        }
        numberOfFreeObjects -= 1;
        return freeObjects[numberOfFreeObjects];
    }

    /**
     * Return object to pool (make it free)
     * @param obj Object to pool.
     */
    @Override
    public void delete(T obj) {
        if (obj == null) throw new NullPointerException("Try to return null to MemoryManager.");
        if (freeObjects == null)
            freeObjects = creator.create(initialCapacity);
        if (numberOfFreeObjects == freeObjects.length) {
            final int capacity = (int) (freeObjects.length * growthFactor);
            final T[] newFreeObjects = creator.create(capacity);
            System.arraycopy(freeObjects, 0, newFreeObjects, 0, freeObjects.length);
            freeObjects = newFreeObjects;
        }
        freeObjects[numberOfFreeObjects] = obj;
        numberOfFreeObjects += 1;
    }
}
